File: lib/mock/liquidations.ts
import type { LiquidationEvent, LiquidationHeatmapCell } from "@/types";

const SYMBOLS = ["BTCUSDT", "ETHUSDT", "SOLUSDT", "LINKUSDT", "AVAXUSDT"];
const EXCHANGES = ["Binance", "OKX", "Bybit", "Bitget"];

export function getMockLiquidations(count = 30): LiquidationEvent[] {
  return Array.from({ length: count }, (_, index) => {
    const symbol = SYMBOLS[index % SYMBOLS.length];
    const side = index % 3 === 0 ? "SHORT" : "LONG";
    const amountUsd = 8_000 + ((index * 2917) % 480_000);
    const basePrice = symbol === "BTCUSDT" ? 64280 : symbol === "ETHUSDT" ? 3412 : symbol === "SOLUSDT" ? 172 : symbol === "LINKUSDT" ? 18.6 : 38.4;

    return {
      id: `liq-${index}`,
      symbol,
      side,
      amountUsd,
      price: Number((basePrice * (1 + ((index % 7) - 3) / 500)).toFixed(2)),
      exchange: EXCHANGES[index % EXCHANGES.length],
      timestamp: new Date(Date.now() - index * 1000 * 60 * 6).toISOString(),
    };
  });
}

export function getMockLiquidationHeatmap(basePrice = 64280, count = 40): LiquidationHeatmapCell[] {
  return Array.from({ length: count }, (_, index) => {
    const offset = (index - count / 2) * (basePrice * 0.0025);
    const distanceFactor = Math.abs(index - count / 2) / (count / 2);
    return {
      price: Number((basePrice + offset).toFixed(2)),
      intensity: Number(Math.max(0.08, 1 - distanceFactor + Math.sin(index) * 0.15).toFixed(2)),
    };
  });
}

export const LIQUIDATION_TOTALS_24H = {
  longUsd: 182_400_000,
  shortUsd: 96_700_000,
};

File: backend/app/api/liquidations.py
"""Real liquidation data, fed by Binance's `forceOrder` WebSocket stream
(see `app/tasks/live_feed.py`'s `_handle_liquidation`) and persisted to
the `liquidations` table.

The heatmap buckets recent liquidations by price into `count` bins around
the current price range — it reflects genuinely observed liquidations, so
it may look sparse until enough events have accumulated (this stream only
started recording once the Data Engine started; there's no historical
liquidation backfill).
"""

from datetime import UTC, datetime

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database.session import get_db
from app.schemas.liquidation import LiquidationEvent, LiquidationHeatmapCell, LiquidationTotals
from app.services.market_repository import get_liquidation_totals_24h, get_recent_liquidations

router = APIRouter(prefix="/api/liquidations", tags=["liquidations"])


def _to_liquidation_out(row) -> LiquidationEvent:
    return LiquidationEvent(
        id=str(row.id),
        symbol=row.symbol,
        side=row.side,
        amount_usd=row.amount_usd,
        price=row.price,
        exchange=row.exchange,
        timestamp=datetime.fromtimestamp(row.timestamp / 1000, tz=UTC).isoformat(),
    )


@router.get("", response_model=list[LiquidationEvent])
async def list_liquidations(
    count: int = Query(30, ge=1, le=200), db: AsyncSession = Depends(get_db)
) -> list[LiquidationEvent]:
    rows = await get_recent_liquidations(db, limit=count)
    return [_to_liquidation_out(row) for row in rows]


@router.get("/totals", response_model=LiquidationTotals)
async def get_totals(db: AsyncSession = Depends(get_db)) -> LiquidationTotals:
    totals = await get_liquidation_totals_24h(db)
    return LiquidationTotals(long_usd=totals["LONG"], short_usd=totals["SHORT"])


@router.get("/heatmap", response_model=list[LiquidationHeatmapCell])
async def get_heatmap(
    symbol: str = "BTCUSDT", count: int = Query(40, ge=1, le=200), db: AsyncSession = Depends(get_db)
) -> list[LiquidationHeatmapCell]:
    rows = await get_recent_liquidations(db, limit=500, symbol=symbol.upper())
    if not rows:
        return []

    prices = [row.price for row in rows]
    low, high = min(prices), max(prices)
    if low == high:
        return [LiquidationHeatmapCell(price=low, intensity=1.0)]

    bucket_width = (high - low) / count
    buckets = [0.0] * count
    for row in rows:
        index = min(count - 1, int((row.price - low) / bucket_width))
        buckets[index] += row.amount_usd

    max_bucket = max(buckets) or 1.0
    return [
        LiquidationHeatmapCell(price=round(low + bucket_width * (i + 0.5), 2), intensity=round(value / max_bucket, 4))
        for i, value in enumerate(buckets)
        if value > 0
    ]
